include/seastar/core/iostream-impl.hh

```diff
@@ -152,26 +153,28 @@ future<> output_stream<CharType>::write(temporary_buffer<CharType> p) noexcept {
 template <typename CharType>
 future<temporary_buffer<CharType>>
 input_stream<CharType>::read_exactly_part(size_t n, tmp_buf out, size_t completed) noexcept {
-    if (available()) {
-        auto now = std::min(n - completed, available());
-        std::copy(_buf.get(), _buf.get() + now, out.get_write() + completed);
-        _buf.trim_front(now);
-        completed += now;
-    }
-    if (completed == n) {
-        return make_ready_future<tmp_buf>(std::move(out));
-    }
-
-    // _buf is now empty
-    return _fd.get().then([this, n, out = std::move(out), completed] (auto buf) mutable {
+    while(true) {
+        size_t avail = available();
+        if (avail) {
+            auto now = std::min(n - completed, avail);
+            std::copy_n(_buf.get(), now, out.get_write() + completed);
+            _buf.trim_front(now);
+            completed += now;
+        }
+    
+        if (completed == n) {
+            co_return out;
+        }
+    
+        // _buf is now empty
+        temporary_buffer<CharType> buf = co_await _fd.get();
```

Comment:
If I understand correctly, even though `co_await` could throw, this will be translated back to an exceptional future,  so that `read_exactly_part` being `noexcept` does not lead to trouble, right?

Reply:
right. `noexcept` is literally `noexcept`, which has nothing to do the coroutine abstraction. the exception "thrown" by `_fd.get()` would be propagated via `promise_type::unhandled_exception()`.